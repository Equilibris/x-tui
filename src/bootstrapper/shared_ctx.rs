mod console;
mod eq;
mod quit;
mod rb;
mod region;
mod test;

pub use console::*;
pub use eq::*;
pub use quit::*;
pub use rb::*;
pub use region::*;
pub use test::*;
