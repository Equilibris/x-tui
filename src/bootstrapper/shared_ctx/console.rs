use std::mem::MaybeUninit;

use chrono::{DateTime, Local, Timelike};
use crossterm::event::KeyCode;
use leptos_reactive::*;
use tui::{
    layout::Rect,
    text::{Span, Spans, Text},
    widgets::{Block, Borders, Clear, Paragraph},
};

use crate::{
    controlflow::{each, show},
    split_word_wrap::split_word_wrap,
};

use super::{EventQueue, Region, RenderBaseAuto};

#[derive(Clone, Copy)]
pub struct Console(RwSignal<Vec<(DateTime<Local>, String)>>);

impl std::ops::Deref for Console {
    type Target = RwSignal<Vec<(DateTime<Local>, String)>>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Console {
    pub fn attach(cx: Scope) -> Self {
        let v = Self(create_rw_signal(
            cx,
            vec![
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
                (
                    Local::now(),
                    "Terminal attached; program is bootstrapped".into(),
                ),
            ],
        ));

        provide_context(cx, v);

        v
    }

    pub fn render_encapsulate(self, cx: Scope, child: impl FnOnce(Scope)) {
        let region: Region = use_context(cx).unwrap();
        let eq: EventQueue = use_context(cx).unwrap();

        let when = create_memo(cx, move |v| {
            let v = v.cloned().unwrap_or(false);
            eq.0.with(move |e| match e {
                crossterm::event::Event::Key(k) if k.code == KeyCode::F(12) => !v,
                _ => v,
            })
        });

        let child_area = Signal::derive(cx, move || {
            let r = region();
            if when() {
                Rect::new(r.x, r.y, r.width, r.height / 2)
            } else {
                r
            }
        });
        let my_area = Signal::derive(cx, move || {
            let r = region();
            let c = child_area();

            Rect::new(0, c.height, r.width, r.height - c.height)
        });

        cx.child_scope(|cx| {
            Region::derive(cx, child_area);

            child(cx);
        });

        show(
            cx,
            when,
            move |cx| {
                let block = store_value(
                    cx,
                    Block::default()
                        .borders(Borders::ALL)
                        .style(tui::style::Style::default().bg(tui::style::Color::Red)),
                );

                let inner_area =
                    Signal::derive(cx, move || block.with(|block| block.inner(my_area())));

                let rb: RenderBaseAuto = use_context(cx).unwrap();

                create_effect(cx, move |_| {
                    let reg = my_area();

                    rb.render(Clear, reg);
                    rb.render(block(), reg);
                });

                each(
                    cx,
                    move || {
                        self.with(move |data| {
                            data.iter()
                                .rev()
                                .take(inner_area().height.into())
                                .cloned()
                                .collect::<Vec<_>>()
                                .into_iter()
                        })
                    },
                    move |_, idx| idx,
                    move |last: Option<&(u16, u16, Paragraph)>, (inst, data), _| {
                        let inner_area = inner_area();

                        let offset = last.map(move |v| v.0 as u16).unwrap_or_default();
                        let available = inner_area.height - offset;

                        if available == 0 {
                            return None;
                        }

                        let data_blocks = split_word_wrap(&data, inner_area.width as usize - 8);

                        let mut lines = vec![];

                        let used_height = std::cmp::min(available.into(), data_blocks.len());

                        let mut data_blocks = data_blocks.into_iter().rev().take(used_height).rev();

                        lines.push(Spans::from(format!(
                            "[{:02}:{:02}] {}",
                            inst.minute(),
                            inst.second(),
                            &data[data_blocks.next().unwrap()]
                        )));

                        for i in data_blocks {
                            lines.push(Spans::from(format!("    >>> {}", &data[i])))
                        }

                        let used_height = used_height as u16;

                        Some((
                            offset + used_height,
                            used_height,
                            Paragraph::new(Text { lines }),
                        ))
                    },
                    move |cx, data| {
                        let rb: RenderBaseAuto = use_context(cx).unwrap();
                        data.with(move |(off, h, p)| {
                            let inner_area = inner_area();

                            let rect = Rect::new(
                                1,
                                inner_area.y + inner_area.height - off,
                                inner_area.width,
                                *h,
                            );

                            rb.render(p.clone(), rect);
                        })
                    },
                );
            },
            |_cx| {},
        );
    }

    pub fn log(&self, msg: impl Into<String>) {
        self.0.update(move |v| v.push((Local::now(), msg.into())))
    }
    pub fn clear(&self) {
        self.0.update(move |v| v.clear())
    }
}
