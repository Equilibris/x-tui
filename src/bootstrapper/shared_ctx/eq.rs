use std::{
    error::Error,
    sync::mpsc::{self},
    time::Duration,
};

use crossterm::event::{self, Event};
use leptos_reactive::*;
use tui::{backend::Backend, layout::Rect};

use super::RenderBase;

#[derive(Clone)]
pub struct EventDispatcher(mpsc::Sender<Event>);

impl EventDispatcher {
    pub fn dispatch(&self, e: Event) -> Result<(), mpsc::SendError<Event>> {
        self.0.send(e)
    }
    pub fn attach(cx: Scope, tx: mpsc::Sender<Event>) {
        provide_context(cx, Self(tx))
    }
}

#[derive(Clone, Copy)]
pub struct EventQueue(pub ReadSignal<Event>);

impl std::ops::Deref for EventQueue {
    type Target = ReadSignal<Event>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl EventQueue {
    pub fn attach(cx: Scope) -> RwSignal<Event> {
        let base = create_rw_signal(cx, Event::FocusGained);
        let v = Self(base.read_only());

        provide_context(cx, v);

        base
    }
    fn dispatch<B: Backend>(
        eq: &RwSignal<Event>,
        e: Event,
        region: &RwSignal<Rect>,
        rb: RenderBase<B>,
    ) -> Result<(), Box<dyn Error>> {
        Ok(match e {
            Event::Resize(h, w) => {
                let rect = tui::layout::Rect::new(0, 0, w, h);
                {
                    let rb = rb.access();
                    let mut rb = rb.lock().unwrap();
                    rb.1.resize(rect);
                    rb.0.resize(rect)?;
                }

                region.set(rect);
                eq.set(Event::Resize(h, w));
            }
            e => eq.set(e),
        })
    }
    pub fn test_poll<B: Backend>(
        eq: &RwSignal<Event>,
        rec: &mpsc::Receiver<Event>,
        region: &RwSignal<Rect>,
        rb: RenderBase<B>,
    ) -> Result<(), Box<dyn Error>> {
        Ok(match rec.recv_timeout(Duration::from_millis(1)) {
            Ok(e) => Self::dispatch(eq, e, region, rb)?,
            Err(e) => match e {
                mpsc::RecvTimeoutError::Timeout => (),
                mpsc::RecvTimeoutError::Disconnected => Err(e)?,
            },
        })
    }
    pub fn poll<B: Backend>(
        eq: &RwSignal<Event>,
        region: &RwSignal<Rect>,
        rb: RenderBase<B>,
    ) -> Result<(), Box<dyn Error>> {
        Ok(if event::poll(Duration::from_millis(10))? {
            Self::dispatch(eq, event::read()?, region, rb)?
        })
    }
}
