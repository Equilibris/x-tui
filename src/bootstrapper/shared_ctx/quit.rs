use leptos_reactive::*;

#[derive(Clone, Copy)]
pub struct Quit(RwSignal<Option<String>>);

impl Quit {
    pub fn attach(cx: Scope) -> Self {
        let v = Self(create_rw_signal(cx, None));
        provide_context(cx, v);
        v
    }

    pub fn quit_with_message(&self, msg: impl Into<String>) {
        self.0.set(Some(msg.into()))
    }
    pub fn quit(&self) {
        self.0.set(Some(String::new()));
    }
    pub fn get_msg(&self) -> String {
        self.0.get().unwrap()
    }
    pub fn should_quit(&self) -> bool {
        self.0.get_untracked().is_some()
    }
}
