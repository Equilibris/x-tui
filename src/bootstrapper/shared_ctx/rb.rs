use std::sync::{Arc, Mutex, Weak};

use leptos_reactive::{provide_context, Scope};
use tui::{backend::Backend, layout::Rect, Terminal};

use crate::bootstrapper::prefix_sum_2d::PrefixSum2d;

// Flushing could be scoped but this is an optimization that has to be evaluated
pub struct RenderBase<B: Backend + 'static>(pub(super) Weak<Mutex<(Terminal<B>, PrefixSum2d)>>);

pub type RenderBaseCT = RenderBase<tui::backend::CrosstermBackend<std::io::Stdout>>;

impl<B: Backend + 'static> RenderBase<B> {
    pub fn attach(cx: Scope, v: Arc<Mutex<(Terminal<B>, PrefixSum2d)>>) {
        let v = Self(Arc::downgrade(&v));
        provide_context(cx, v)
    }

    pub fn access(&self) -> Arc<Mutex<(Terminal<B>, PrefixSum2d)>> {
        self.0.upgrade().unwrap()
    }
    pub fn render<T: tui::widgets::Widget>(&self, widget: T, area: Rect) {
        let upgrade = self.access();
        let mut term = upgrade.try_lock().unwrap();

        term.0.get_frame().render_widget(widget, area);
        term.1.insert(area);
    }

    pub fn do_frame(&self) -> Result<(), std::io::Error> {
        let term = self.access();
        let mut term = &mut term.try_lock().unwrap();

        let sz = term.0.size()?;
        let data = term.0.current_buffer_mut().clone();
        let base_data = term.1.clone();

        let data = data
            .content
            .iter()
            .zip(base_data.iter())
            .filter_map(|(cell, (x, y, v))| (v > 0).then_some((x, y, cell)));

        term.0.backend_mut().draw(data)?;
        term.0.backend_mut().flush()?;
        term.1.clear();

        Ok(())
    }
}

impl<B: Backend> Clone for RenderBase<B> {
    fn clone(&self) -> Self {
        Self(Weak::clone(&self.0))
    }
}
