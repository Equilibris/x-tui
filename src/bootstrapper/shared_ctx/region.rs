use leptos_reactive::{create_rw_signal, provide_context, ReadSignal, RwSignal, Scope};
use tui::layout::Rect;

#[derive(Clone, Copy)]
pub struct Region(pub ReadSignal<Rect>);

impl Region {
    pub fn attach(cx: Scope, v: Rect) -> RwSignal<Rect> {
        let base = create_rw_signal(cx, v);
        let v = Self(base.read_only());

        provide_context(cx, v);

        base
    }
}
