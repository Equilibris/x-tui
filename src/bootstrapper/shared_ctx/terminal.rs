use crossterm::event::KeyCode;
use leptos_reactive::{
    create_effect, create_memo, create_rw_signal, provide_context, use_context, RwSignal, Scope,
    Signal,
};
use tui::{
    layout::Rect,
    widgets::{Block, Borders},
};

use crate::controlflow::show;

use super::{EventQueue, Region, RenderBaseAuto};

#[derive(Clone, Copy)]
pub struct Terminal(RwSignal<Vec<String>>);

impl Terminal {
    pub fn attach(cx: Scope) -> Self {
        let v = Self(create_rw_signal(cx, vec![]));

        provide_context(cx, v);

        v
    }

    pub fn render_encapsulate(self, cx: Scope, child: impl FnOnce(Scope)) {
        let region: Region = use_context(cx).unwrap();
        let eq: EventQueue = use_context(cx).unwrap();

        let when = create_memo(cx, move |v| {
            let v = v.cloned().unwrap_or(false);
            eq.0.with(move |e| match e {
                crossterm::event::Event::Key(k) if k.code == KeyCode::F(12) => !v,
                _ => v,
            })
        });

        let child_area = Signal::derive(cx, move || {
            let r = region();
            if when() {
                Rect::new(r.x, r.y, r.width, r.height / 2)
            } else {
                r
            }
        });
        let my_area = Signal::derive(cx, move || {
            let r = region();
            let c = region();

            Rect::new(0, c.height + 1, r.width, r.height - c.height)
        });

        cx.child_scope(|cx| {
            Region::derive(cx, child_area);

            child(cx);
        });

        show(
            cx,
            when,
            move |cx| {
                // let quit: Quit = use_context(cx).unwrap();

                let rb: RenderBaseAuto = use_context(cx).unwrap();

                create_effect(cx, move |_| {
                    let reg = my_area();

                    let block = Block::default()
                        .borders(Borders::ALL)
                        .style(tui::style::Style::default().bg(tui::style::Color::Red));

                    rb.render(block, reg);
                });
            },
            |_cx| {},
        );
    }

    pub fn enqueue(&self, msg: String) {
        self.0.update(move |v| v.push(msg))
    }
    pub fn clear(&self) {
        self.0.update(move |v| v.clear())
    }
}
