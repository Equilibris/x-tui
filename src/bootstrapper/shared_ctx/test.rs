use leptos_reactive::{create_rw_signal, provide_context, ReadSignal, Scope, WriteSignal};

#[derive(Clone, Copy)]
pub struct RenderCounter(pub ReadSignal<usize>);

impl RenderCounter {
    pub fn attach(cx: Scope) -> WriteSignal<usize> {
        let (r, w) = create_rw_signal(cx, 0).split();

        let v = Self(r);

        provide_context(cx, v);

        w
    }
}
