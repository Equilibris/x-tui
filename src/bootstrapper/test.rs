use leptos_reactive::Scope;
use std::{
    error::Error,
    sync::{Arc, Mutex},
};
use tui::{backend::TestBackend, Terminal};

use super::{bootstrap, init::TestInit, prefix_sum_2d::PrefixSum2d};

pub fn test_bootstrap(
    boot: impl FnOnce(Scope) + 'static,
    w: u16,
    h: u16,
    once: bool,
) -> Result<(), Box<dyn Error>> {
    let backend = TestBackend::new(w, h);
    let terminal = Terminal::new(backend)?;
    let ps = PrefixSum2d::new(terminal.size()?);

    let terminal = Arc::new(Mutex::new((terminal, ps)));

    let init = TestInit::new(once);

    bootstrap(terminal, boot, init)?;
    Ok(())
}

#[cfg(test)]
use super::shared_ctx::RenderBase;
#[cfg(test)]
pub fn assert_rb(rb: &RenderBase<TestBackend>, snap_name: &'_ str) {
    let term = rb.access();
    let mut term = term.try_lock().unwrap();

    let rect = term.0.size().unwrap();

    let mut symbols = term
        .0
        .current_buffer_mut()
        .content
        .iter()
        .map(|v| v.symbol.as_str());
    let out = (0..rect.height)
        .map(|_| {
            (0..rect.width)
                .map(|_| symbols.next().unwrap())
                .collect::<String>()
        })
        .intersperse("$\n".into())
        .chain(std::iter::once("$".into()))
        .collect::<String>();

    insta::assert_snapshot!(snap_name, out.as_str());
}
