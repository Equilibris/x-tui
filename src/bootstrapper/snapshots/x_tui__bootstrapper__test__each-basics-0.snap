---
source: src/bootstrapper/test.rs
expression: out.as_str()
---
0         $
1         $
2         $
3         $
4         $
          $
          $
          $
          $
          $
